[CalDav] Fails for calendar resources containing curly braces

The error I get is below. My entry from openhab.cfg is below as well.

################################caldav
caldavio:jason:reloadInterval=1
caldavio:jason:preloadTime=2880
caldavio:jason:oauth=true
caldavio.jason:disableCertificateVerification=false

caldavio:jason:username=<email omitted>@gmail.com
caldavio:jason:password=<password omitted>
caldavio:jason:url=https://www.google.com/calendar/dav/<email omitted>@gmail.com/events
caldavPersonal:usedCalendars=jason

2017-08-04 10:08:27.697 [ERROR] [o.o.i.c.i.job.EventReloaderJob] - error while loading calendar entries: Illegal character in path at index 65: https://www.google.com/calendar/dav/jason.brunk@gmail.com/events/{4913D9DD-6C56-4445-97FE-623E60533DB1}.ics
java.lang.IllegalArgumentException: Illegal character in path at index 65: https://www.google.com/calendar/dav/jason.brunk@gmail.com/events/{4913D9DD-6C56-4445-97FE-623E60533DB1}.ics
at java.net.URI.create(URI.java:852) ~[na:1.8.0_131]
at org.apache.http.client.methods.HttpGet.(HttpGet.java:69) ~[na:na]
at com.github.sardine.impl.SardineImpl.get(SardineImpl.java:745) ~[na:na]
at com.github.sardine.impl.SardineImpl.get(SardineImpl.java:741) ~[na:na]
at com.github.sardine.impl.SardineImpl.get(SardineImpl.java:732) ~[na:na]
at com.github.sardine.impl.SardineImpl.get(SardineImpl.java:145) ~[na:na]
at org.openhab.io.caldav.internal.job.EventReloaderJob.loadEvents(EventReloaderJob.java:314) ~[na:na]
at org.openhab.io.caldav.internal.job.EventReloaderJob.execute(EventReloaderJob.java:135) ~[na:na]
at org.quartz.core.JobRunShell.run(JobRunShell.java:213) [quartz-all-2.1.7.jar:na]
at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557) [quartz-all-2.1.7.jar:na]
Caused by: java.net.URISyntaxException: Illegal character in path at index 65: https://www.google.com/calendar/dav/jason.brunk@gmail.com/events/{4913D9DD-6C56-4445-97FE-623E60533DB1}.ics
at java.net.URI$Parser.fail(URI.java:2848) ~[na:1.8.0_131]
at java.net.URI$Parser.checkChars(URI.java:3021) ~[na:1.8.0_131]
at java.net.URI$Parser.parseHierarchical(URI.java:3105) ~[na:1.8.0_131]
at java.net.URI$Parser.parse(URI.java:3053) ~[na:1.8.0_131]
at java.net.URI.(URI.java:588) ~[na:1.8.0_131]
at java.net.URI.create(URI.java:850) ~[na:1.8.0_131]
… 9 common frames omitted

Looks like something to do with the { }

--- a/bundles/io/org.openhab.io.caldav/src/main/java/org/openhab/io/caldav/internal/CalDavLoaderImpl.java
+++ b/bundles/io/org.openhab.io.caldav/src/main/java/org/openhab/io/caldav/internal/CalDavLoaderImpl.java
@@ -113,8 +113,8 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                 scheduler = new StdSchedulerFactory().getScheduler();
                 this.removeAllJobs();
             } catch (SchedulerException e) {
-                log.error("cannot get job-scheduler", e);
-                throw new IllegalStateException("cannot get job-scheduler", e);
+                log.warn("Cannot get job scheduler", e);
+                throw new IllegalStateException("Cannot get job scheduler", e);
             }
 
             this.startLoading();
@@ -136,7 +136,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
         try {
             this.removeAllJobs();
         } catch (SchedulerException e) {
-            log.error("cannot remove jobs: " + e.getMessage(), e);
+            log.warn("Cannot remove jobs: {}", e.getMessage(), e);
         }
     }
 
@@ -288,7 +288,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                     try {
                         this.scheduler.deleteJob(JobKey.jobKey(timerKey));
                     } catch (SchedulerException e) {
-                        log.error("cannot cancel event with job-id: " + timerKey, e);
+                        log.warn("Cannot cancel event with job id: {}", timerKey, e);
                     }
                 }
                 eventContainerOld.getTimerMap().clear();
@@ -302,7 +302,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                         try {
                             notifier.eventRemoved(event);
                         } catch (Exception e) {
-                            log.error("error while invoking listener", e);
+                            log.warn("Error while invoking listener", e);
                         }
                     }
                 }
@@ -312,7 +312,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                         try {
                             notifier.eventLoaded(event);
                         } catch (Exception e) {
-                            log.error("error while invoking listener", e);
+                            log.warn("Error while invoking listener", e);
                         }
                     }
                 }
@@ -324,7 +324,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                             try {
                                 createJob(eventContainer, event, index);
                             } catch (SchedulerException e) {
-                                log.error("cannot create jobs for event '{}': ", event.getShortName(), e.getMessage());
+                                log.warn("Cannot create jobs for event '{}': {}", event.getShortName(), e.getMessage());
                             }
                         }
                         index++;
@@ -363,16 +363,16 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                             try {
                                 notifier.eventLoaded(event);
                             } catch (Exception e) {
-                                log.error("error while invoking listener", e);
+                                log.warn("Error while invoking listener", e);
                             }
                         }
                         if (createTimer) {
                             try {
-                                log.trace("creating job for event {} ", event.getShortName());
+                                log.trace("creating job for event {}", event.getShortName());
                                 createJob(eventContainerOld, event, eventContainerOld.getEventList().size() - 1);
                                 // -1 because we already added the event to eventContainerOld
                             } catch (SchedulerException e) {
-                                log.error("cannot create jobs for event '{}': ", event.getShortName(), e.getMessage());
+                                log.warn("Cannot create jobs for event '{}': {}", event.getShortName(), e.getMessage());
                             }
                         }
                     }
@@ -390,7 +390,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                     try {
                         notifier.eventLoaded(event);
                     } catch (Exception e) {
-                        log.error("error while invoking listener", e);
+                        log.warn("Error while invoking listener", e);
                     }
                 }
             }
@@ -401,7 +401,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                         try {
                             createJob(eventContainer, event, index);
                         } catch (SchedulerException e) {
-                            log.error("cannot create jobs for event: " + event.getShortName());
+                            log.warn("Cannot create jobs for event: {}", event.getShortName());
                         }
                     }
                     index++;
@@ -476,7 +476,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                 this.scheduler.scheduleJob(jobTrigger);
                 log.info("reload job scheduled for: {}", eventRuntime.getConfig().getKey());
             } catch (SchedulerException e) {
-                log.error("cannot schedule calendar-reloader", e);
+                log.warn("Cannot schedule calendar reloader", e);
             }
             // next event 10 seconds later
             i += 10;
@@ -505,10 +505,10 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                 .get(calDavEvent.getCalendarId());
         CalDavConfig config = calendarRuntime.getConfig();
         if (config == null) {
-            log.error("cannot find config for calendar id: {}", calDavEvent.getCalendarId());
+            log.warn("Cannot find config for calendar id: {}", calDavEvent.getCalendarId());
+            return;
         }
         Sardine sardine = Util.getConnection(config);
-
         Calendar calendar = Util.createCalendar(calDavEvent, defaultTimeZone);
 
         try {
@@ -518,7 +518,7 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
                 try {
                     sardine.delete(fullIcsFile);
                 } catch (IOException e) {
-                    log.error("cannot remove old ics file: {}", fullIcsFile);
+                    log.warn("Cannot remove old ics file: {}", fullIcsFile);
                 }
             } else {
                 log.debug("event is new: {}", fullIcsFile);
@@ -533,9 +533,9 @@ public class CalDavLoaderImpl extends AbstractActiveService implements ManagedSe
             eventContainer.setLastChanged(calDavEvent.getLastChanged());
             this.addEventToMap(eventContainer, false);
         } catch (UnsupportedEncodingException e) {
-            log.error("cannot write event", e);
+            log.warn("Cannot write event", e);
         } catch (IOException e) {
-            log.error("cannot write event", e);
+            log.warn("Cannot write event", e);
         }
     }
 

--- a/bundles/io/org.openhab.io.caldav/src/main/java/org/openhab/io/caldav/internal/job/EventReloaderJob.java
+++ b/bundles/io/org.openhab.io.caldav/src/main/java/org/openhab/io/caldav/internal/job/EventReloaderJob.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016 by the respective copyright holders.
+ * Copyright (c) 2010-2017 by the respective copyright holders.
  *
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
@@ -14,6 +14,8 @@ import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -99,22 +101,22 @@ public class EventReloaderJob implements Job {
                                     new org.joda.time.DateTime(icsFile.lastModified()), fis, eventRuntime.getConfig(),
                                     new ArrayList<String>(), true);
                         } catch (IOException e) {
-                            log.error("cannot load events for file: " + icsFile, e);
+                            log.warn("Cannot load events for file: {}", icsFile, e);
                         } catch (ParserException e) {
-                            log.error("cannot load events for file: " + icsFile, e);
+                            log.warn("Cannot load events for file: {}", icsFile, e);
                         }
                     }
                     break;
                 }
             } catch (Throwable e) {
-                log.error("cannot load events", e);
+                log.warn("Cannot load events", e);
             } finally {
                 cachedEventsLoaded.put(config, true);
             }
         }
 
         try {
-            log.debug("loading events for config: " + config);
+            log.debug("loading events for config: {}", config);
             List<String> oldEventIds = new ArrayList<String>();
             for (EventContainer eventContainer : eventRuntime.getEventMap().values()) {
                 oldEventIds.add(eventContainer.getFilename());
@@ -126,7 +128,7 @@ public class EventReloaderJob implements Job {
                 if (log.isDebugEnabled()) {
                     for (int i = 0; i < eventContainer.getEventList().size(); i++) {
                         CalDavEvent elem = eventContainer.getEventList().get(i);
-                        log.debug("old eventlist contient l'evenement : {} -- deb : {} -- fin : {} -- lastchang {}",
+                        log.debug("old eventlist contains the event : {} -- deb : {} -- fin : {} -- lastchanged {}",
                                 elem.getName(), elem.getStart(), elem.getEnd(), elem.getLastChanged());
                     }
                 }
@@ -140,11 +142,9 @@ public class EventReloaderJob implements Job {
                 try {
                     notifier.calendarReloaded(config);
                 } catch (Exception e) {
-                    log.error("error while invoking listener", e);
+                    log.warn("Error while invoking listener", e);
                 }
             }
-
-            // printAllEvents();
 
             // print All scheduled jobs :
             if (log.isDebugEnabled()) {
@@ -169,12 +169,12 @@ public class EventReloaderJob implements Job {
             }
 
         } catch (SardineException e) {
-            log.error("error while loading calendar entries: {} ({} - {} )", e.getMessage(), e.getStatusCode(),
+            log.warn("Sardine error while loading calendar entries: {} ({} - {})", e.getMessage(), e.getStatusCode(),
                     e.getResponsePhrase(), e);
-            throw new JobExecutionException("error while loading calendar entries", e, false);
+            throw new JobExecutionException("Sardine error while loading calendar entries", e, false);
         } catch (Exception e) {
-            log.error("error while loading calendar entries: {}", e.getMessage(), e);
-            throw new JobExecutionException("error while loading calendar entries", e, false);
+            log.warn("Error while loading calendar entries: {}", e.getMessage(), e);
+            throw new JobExecutionException("Error while loading calendar entries", e, false);
         }
     }
 
@@ -184,7 +184,7 @@ public class EventReloaderJob implements Job {
         for (String filename : oldMap) {
             EventContainer eventContainer = eventRuntime.getEventContainerByFilename(filename);
             if (eventContainer == null) {
-                log.error("cannot find event container for filename: {}", filename);
+                log.warn("Cannot find event container for filename: {}", filename);
                 continue;
             }
 
@@ -202,7 +202,7 @@ public class EventReloaderJob implements Job {
                     boolean deleteJob = CalDavLoaderImpl.instance.getScheduler().deleteJob(JobKey.jobKey(jobId, group));
                     log.debug("old job ({}) deleted? {}", jobId, deleteJob);
                 } catch (SchedulerException e) {
-                    log.error("cannot delete job '{}'", jobId);
+                    log.warn("Cannot delete job '{}'", jobId, e);
                 }
             }
             eventContainer.getTimerMap().clear();
@@ -212,18 +212,16 @@ public class EventReloaderJob implements Job {
                     try {
                         notifier.eventRemoved(event);
                     } catch (Exception e) {
-                        log.error("error while invoking listener", e);
+                        log.warn("Error while invoking listener", e);
                     }
                 }
             }
 
             ConcurrentHashMap<String, EventContainer> eventContainerMap = eventRuntime.getEventMap();
-            if (eventContainer != null) {
-                this.removeFromDisk(eventContainer);
+            this.removeFromDisk(eventContainer);
 
-                log.debug("remove deleted event: {}", eventContainer.getEventId());
-                eventContainerMap.remove(eventContainer.getEventId());
-            }
+            log.debug("remove deleted event: {}", eventContainer.getEventId());
+            eventContainerMap.remove(eventContainer.getEventId());
         }
     }
 
@@ -247,7 +245,7 @@ public class EventReloaderJob implements Job {
 
         List<DavResource> list = sardine.list(config.getUrl(), 1, false);
 
-        log.trace("before load events : oldeventsid contains : {}", oldEventIds.toString());
+        log.trace("before load events : oldeventids contains : {}", oldEventIds);
 
         for (DavResource resource : list) {
             final String filename = Util.getFilename(resource.getName());
@@ -262,7 +260,7 @@ public class EventReloaderJob implements Job {
                 // ==>we must not have one of them remaining in oldEventIds var (bad chosen name, cause it's a list of
                 // oldEventContainers's filename, so with doubles possible)
                 // or the remaining jobs with this filename will get unscheduled on the "removeDeletedEvents(config,
-                // oldEventIds)" call (line 136)
+                // oldEventIds)" call
                 oldEventIds.removeAll(Arrays.asList(filename));
 
                 // must not be loaded
@@ -281,14 +279,14 @@ public class EventReloaderJob implements Job {
                                 .isAfter(org.joda.time.DateTime.now().plusMinutes(config.getReloadMinutes()))) {
                             // the event is calculated as long as the next reload
                             // interval can handle this
-                            log.trace("skipping resource {}, not changed (calculated until: {})", resource.getName(),
+                            log.trace("skipping resource {}; not changed. calculated until: {}", resource.getName(),
                                     eventContainer.getCalculatedUntil());
                             continue;
                         }
 
                         if (eventContainer.isHistoricEvent()) {
                             // no more upcoming events, do nothing
-                            log.trace("skipping resource {}, not changed (historic)", resource.getName());
+                            log.trace("skipping resource {}; not changed (historic)", resource.getName());
                             continue;
                         }
 
@@ -306,22 +304,28 @@ public class EventReloaderJob implements Job {
 
                 // prepare resource url
                 URL url = new URL(config.getUrl());
+                log.trace("Raw URL: {}", url);
                 String resourcePath = resource.getPath();
-                String escapedResource = resource.getName().replaceAll("/", "%2F");
-                resourcePath = resourcePath.replace(resource.getName(), escapedResource);
-                url = new URL(url.getProtocol(), url.getHost(), url.getPort(), resourcePath);
-
-                InputStream inputStream = sardine.get(url.toString().replaceAll(" ", "%20"));
+                URI uri;
+                try {
+                    uri = new URI(url.getProtocol(), null, url.getHost(), url.getPort(), resourcePath, null, null);
+                } catch(URISyntaxException e) {
+                    log.warn("Unable to load events; the provided string could not be parsed as a valid URI: '{}'", url);
+                    return;
+                }
+                url = uri.toURL();
+                log.trace("URL after encoding: {}", url);
+                InputStream inputStream = sardine.get(url.toString());
 
                 this.loadEvents(filename, lastResourceChangeFS, inputStream, config, oldEventIds, false);
             } catch (ParserException e) {
-                log.error("error parsing ics file: " + filename, e);
+                log.warn("Parser exception parsing ics file: {}", filename, e);
             } catch (SardineException e) {
-                log.error("error reading ics file: " + filename, e);
+                log.warn("Sardine exception reading ics file: {}", filename, e);
             }
         }
 
-        log.trace("after load events : oldeventsid contains : {}", oldEventIds.toString());
+        log.trace("after load events : oldeventids contains : {}", oldEventIds.toString());
     }
 
     public void loadEvents(String filename, org.joda.time.DateTime lastResourceChangeFS, final InputStream inputStream,
@@ -334,7 +338,6 @@ public class EventReloaderJob implements Job {
         final UnfoldingReader uin = new UnfoldingReader(in, 50, true);
         Calendar calendar = builder.build(uin);
         uin.close();
-        // log.trace("calendar: {}", calendar);
 
         EventContainer eventContainer = new EventContainer(config.getKey());
         eventContainer.setFilename(filename);
@@ -347,7 +350,7 @@ public class EventReloaderJob implements Job {
 
         final ComponentList<CalendarComponent> vEventComponents = calendar.getComponents(Component.VEVENT);
         if (vEventComponents.size() == 0) {
-            log.debug("could not find a VEVENT from calendar build, based on file {}", filename);
+            log.debug("could not find a VEVENT from calendar build based on file {}", filename);
             // no events inside
             if (!readFromFile) {
                 Util.storeToDisk(config.getKey(), filename, calendar);
@@ -371,7 +374,7 @@ public class EventReloaderJob implements Job {
                 if (lastModifedVEventOverAll == null || lastModifedVEvent.isAfter(lastModifedVEventOverAll)) {
                     lastModifedVEventOverAll = lastModifedVEvent;
                 }
-                if (eventContainer != null && !lastModifedVEvent.isBefore(eventContainer.getLastChanged())) {
+                if (!lastModifedVEvent.isBefore(eventContainer.getLastChanged())) {
                     // check if some timers or single (from repeating events) have
                     // to be created
                     if (eventContainer.getCalculatedUntil() != null && vEventComponents.size() == 1

